core/src/main/scala/cats/Selective.scala

```diff
@@ -0,0 +1,7 @@
+package cats
+
+import simulacrum.{typeclass}
+import scala.annotation.implicitNotFound
+
+@implicitNotFound("Could not find an instance of Selective for ${F}")
+@typeclass trait Selective[F[_]] extends Applicative[F] with Select[F]
```

Comment:
I haven't implemented any derived methods yet.  Some will go here, some can go in `Select`.

Reply:
I'm -1 on Select.

Reply:
I agree.  Some more methods derived from `select` may end up on `Apply` then.  We'll see.

Reply:
Started pulling on the string to prove this was so. `ifS` can be expressed with `apply`.  `whenS` needs `Applicative` so I put it on `Selective`, but why not `Applicative`?  And then what's left of `Selective`?

Maybe the `Selective` typeclass is really `RigidSelective`, which adds the `apS == <$>` law.  All monads are rigidly selective, but maybe not all rigid selectives are monads.

Reply:
I am +1 on this. I maybe wouldn't put select on Apply in that case. I think weak law typeclasses aren't that interesting. I also don't like it when you have something that may or may not be true.

If non rigid selective can be implemented with Apply, in my view, it isn't a new typeclass.

Reply:
Relevant discussion from [the original paper](https://dl.acm.org/ft_gateway.cfm?id=3341694).

> Such loose requirements on select with respect to unnecessary effects might seem troublesome.
Below we list three reasons why we chose to keep the requirements loose.
• Requiring unnecessary effects to be skipped rules out any instances that could be used for
static analysis. Indeed, the only way to obey this law would be to look at actual values at
runtime, since it is impossible to statically know if f (Either a b) contains a Right value.
• One might suggest having two classes Selective and StrictSelective, the latter with
stricter requirements. But this second class would be useless: to statically analyse a computation you would have to express it via Selective, since it is inhabited by instances like Over.
As for the execution, a Monad is perfectly suitable, as we will see in ğ5.2.
• Finally, there is a good precedent: the Applicative type class has no requirements on the
order in which effects are executed: left-to-right, right-to-left, or in parallel. This loose
specification allows some instances to execute effects sequentially (typically from left to
right), and other instances to execute effects in parallel. Note that as soon as we also have a
Monad, we gain an additional requirement (<*>) = ap, which tells us that, at the semantic
level, the result should be as if the effects were executed in sequence from left to right. We
follow the same approach by requiring select = selectM if f is also a Monad (see below

* I think `select` does belong somewhere below `RigidSelective`, for the static analysis reasons.
* `RigidSelective` could still be a fine constraint if there are rigid selectives that aren't also monads, but it may be more obscure than `FlatMap`.
* If all `Apply` are `Selective`, the only value in a `Selective` constraint is providing extended operations that an `Apply` promises not to do.   `F: Applicative` means "I know my conditionals will evaluate both branches." and `F: Selective` says "eh, maybe not."  Is that parametricity worth it?

Reply:
I am thinking of parsing, or IO use cases where "maybe not" would totally break things.

I don't quite see why static analysis is impossible if some effects are skipped, you still know it is one of two cases. Yes it removes branches, but only by saying both run, how is that different that knowing that there are some nodes that can be labeled to possibly be unneeded.

I'm confused why Selective exists at all if all Apply are lawful Selectives. That seems very strange to me. I can't think of another typeclass for which that's true. It is true some things aren't totally locked down by laws (as suggested), but I don't recall a case where a subclass is enough to implement the superclass. That just seems wrong to me.

So, I guess it is okay to put select on Apply, and derived functions, but I would want the law on RigidSelective (which I would consider just calling Selective in cats) that in select you definitely don't evaluate the right arg when the Either is Right.

Reply:
> I don't quite see why static analysis is impossible if some effects are skipped,

I don't think this is the claim.  I think the claim is that different static analyses can be performed by skipping or not.  `Over` records all possible effects, and `Under` records all guaranteed effects.  Establishing these bounds is possible because of this vagary in the law.

> I'm confused why Selective exists at all if all Apply are lawful Selectives. 

I'm still forming my thoughts on this.  The authors hang their hats on this:

> Some applicative functors are not monads, e.g. the Const
functor (ğ2.2), but every applicative functor is also a selective functor, as witnessed by the function
selectA. The subclass relationship Applicative f => Selective f is justified only by the extra
method select in Selective. While select = selectA is a valid implementation of select, it is
not the only useful implementation,

I can't think of another typeclass in Cats that can be lawfully derived from a subclass, but is there one where we would override a default implementation to change behavior, rather than optimize?

Reply:
Taking a straight path up the hierarchy, almost every subclass introduces exactly one abstract method.  `Monad` and `Async` are two exceptions, but in each case, the relationship between the two abstract methods is legislated.  `Concurrent` does not have a snappy law relating its two abstract operations, but has laws that utilize both.

* `Invariant` -> `imap`
* `Functor` -> `map`
* `Apply` -> `ap`
* `Applicative` -> `pure`
* `Monad` -> `flatMap`, `tailRecM`
* `Bracket` -> `bracketCase`
* `Sync` -> `suspend`
* `Async` -> `async`, `asyncF`
* `Concurrent` -> `racePair`, `start`
* `ConcurrentEffect` -> `runCancelable`

If we put `select` on `Apply` as an abstract method, we'd violate the principle of all abstract methods on a typeclass being legally related.  If we put it with a default implementation equivalent to `selectA`, we'd be encouraging overrides that change behavior, which goes against your principle of laws checking consistency with default behavior.  We get around both of these with a `Selective` typeclass.  

I'm back to leaning toward this being a real typeclass.

Reply:
`Selective` is out, but its spirit lives on.